# 4.9.1 Incorporate customer data set

Table of Contents

#### 1. Import data and libraries
#### 2. Data wrangling
#### 3. Data consistency checks
#### 4. Merge dataframes
#### 5. Export final dataframe

## 01. Import data and libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create a path

path = r'/Users/giadairene/Documents/CareerFoundry Data Analytics/Data Analytics Immersion/Achievement 4/Instacart Basket Analysis'

In [3]:
# Import customer data set

df_customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## 02. Data wrangling

In [4]:
# Take a glance at the dataframe

df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
df_customer.shape

(206209, 10)

### Renaming columns

In [6]:
df_customer.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [8]:
df_customer.rename(columns = {'surname' : 'last_name'}, inplace = True)

In [9]:
df_customer.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [10]:
df_customer.rename(columns = {'STATE' : 'state'}, inplace = True)

In [11]:
df_customer.rename(columns = {'Age' : 'age'}, inplace = True)

In [12]:
df_customer.rename(columns = {'n_dependants' : 'no_of_dependants'}, inplace = True)

In [13]:
df_customer.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [14]:
# Check output

df_customer.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependants family_status  income  
0                 3       married  165665  
1                 0        single   59285  
2                 2       married   99568  
3                 0        single   42049  
4                 1       married   40374

### Changing variables datatype

In [15]:
# Check variables datatype for different columns

df_customer['user_id'].dtype

dtype('int64')

In [16]:
df_customer['first_name'].dtype

dtype('O')

In [17]:
df_customer['last_name'].dtype

dtype('O')

In [18]:
df_customer['gender'].dtype

dtype('O')

In [19]:
df_customer['state'].dtype

dtype('O')

In [20]:
df_customer['age'].dtype

dtype('int64')

In [21]:
df_customer['date_joined'].dtype

dtype('O')

In [28]:
df_customer[['user_id', 'date_joined']][df_customer['date_joined']!='1/1/2017']

user_id date_joined
159      51556    1/2/2017
160      59989    1/2/2017
161     183738    1/2/2017
162     190771    1/2/2017
163      11866    1/2/2017
...        ...         ...
206204  168073    4/1/2020
206205   49635    4/1/2020
206206  135902    4/1/2020
206207   81095    4/1/2020
206208   80148    4/1/2020

[206050 rows x 2 columns]

In [22]:
df_customer['no_of_dependants'].dtype

dtype('int64')

In [23]:
df_customer['family_status'].dtype

dtype('O')

In [24]:
df_customer['income'].dtype

dtype('int64')

In [25]:
# Change variables datatype

df_customer['user_id'] = df_customer['user_id'].astype('str')

In [27]:
# Check the output

df_customer['user_id'].dtype

dtype('O')

#### All the other variables types look fine to me.

## 03. Data consistency checks

In [29]:
# Check basic stats

df_customer.describe()

age  no_of_dependants         income
count  206209.000000     206209.000000  206209.000000
mean       49.501646          1.499823   94632.852548
std        18.480962          1.118433   42473.786988
min        18.000000          0.000000   25903.000000
25%        33.000000          0.000000   59874.000000
50%        49.000000          1.000000   93547.000000
75%        66.000000          3.000000  124244.000000
max        81.000000          3.000000  593901.000000

#### The max income value looks a bit odd, should be checked with the client.

### Missing values

In [30]:
# Check for missing values

df_customer.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
family_status           0
income                  0
dtype: int64

In [32]:
# Create a subset containing only missing values

df_nan = df_customer[df_customer['first_name'].isnull() == True]

In [33]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependants family_status  income  
53                     2       married   41709  
73                     0        single   82518  
82                     3       married  117099  
99                     1       married  155673  
105                    3       married  151819  
...                  ...           ...     ...  
206038                 3       married   87783  
206044                 2       married  108488  
206090                 1       married   45275  
206162                 0        single   56325  
206171                 2       married   59222  

[11259 rows x 10 columns]

#### The missing values in the 'first_name' column will not affect the analysis, therefore no action is required.

### Duplicate values

In [34]:
# Looking for full duplicates within the dataframe

df_dups = df_customer[df_customer.duplicated()]

In [35]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependants, family_status, income]
Index: []

#### No duplicates found.

## 04. Merge dataframes

In [36]:
# Import remaining Instacart data

ords_prods_grouped = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_grouped.pkl'))

In [37]:
# Check dataframe

ords_prods_grouped.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc     busiest_day  busiest_period_of_day  \
0  ...    both  Mid-range product  Regularly busy            Most orders   
1  ...    both  Mid-range product  Regularly busy         Average orders   
2  ...    both  Mid-range product    Busiest days         Average orders   
3  ...    both  Mid-range product      Least busy            Most orders   
4  ...    both  Mid-range product      Least busy         Average orders   

  max_order      loyalty_flag average_product_price spending_flag  \
0        32  Regular customer              6.935811   Low spender   
1        32  Regular customer              6.935811   Low spender   
2         5      New customer              7.930208   Low spender   
3         3      New customer              4.972414   Low spender   
4         3      New customer              4.972414   Low spender   

   median_days_between_orders     frequency_flag  
0                         8.0  Frequent customer  
1                         8.0  Frequent customer  
2                         8.0  Frequent customer  
3                         9.0  Frequent customer  
4                         9.0  Frequent customer  

[5 rows x 24 columns]

In [39]:
# Check datatype of ords_prods_grouped['user_id'] column

ords_prods_grouped['user_id'].dtype

dtype('int64')

In [40]:
# Change datatype to string

ords_prods_grouped['user_id'] = ords_prods_grouped['user_id'].astype('str')

In [41]:
# Merge dataframes

df_final = df_customer.merge(ords_prods_grouped, on = 'user_id', indicator = 'merge_flag')

In [42]:
# Check the output

df_final.head()

user_id first_name last_name  gender     state  age date_joined  \
0   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4   26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   no_of_dependants family_status  income  ...    price_range_loc  \
0                 3       married  165665  ...  Mid-range product   
1                 3       married  165665  ...  Mid-range product   
2                 3       married  165665  ...  Mid-range product   
3                 3       married  165665  ...  Low-range product   
4                 3       married  165665  ...  Low-range product   

      busiest_day  busiest_period_of_day  max_order  loyalty_flag  \
0    Busiest days            Most orders          8  New customer   
1  Regularly busy            Most orders          8  New customer   
2    Busiest days            Most orders          8  New customer   
3  Regularly busy            Most orders          8  New customer   
4      Least busy            Most orders          8  New customer   

   average_product_price  spending_flag  median_days_between_orders  \
0               7.988889    Low spender                        19.0   
1               7.988889    Low spender                        19.0   
2               7.988889    Low spender                        19.0   
3               7.988889    Low spender                        19.0   
4               7.988889    Low spender                        19.0   

     frequency_flag  merge_flag  
0  Regular customer        both  
1  Regular customer        both  
2  Regular customer        both  
3  Regular customer        both  
4  Regular customer        both  

[5 rows x 34 columns]

In [44]:
# Check whether there was a full match between the two dataframes

df_final['merge_flag'].value_counts()

merge_flag
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The resulting dataframe (after the merge) has 32,404,859 rows, and each of those rows have information found in both input dataframes, as we used an inner join for the purposes of this project.

## 05. Export final dataframe

In [45]:
# Export data to pkl

df_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'final.pkl'))